File: apps/mobile/src/lib/viewport-tiles.ts
/**
 * Quantizes what the user sees into fixed 0.02° tiles.
 *
 * Tiles rather than the raw viewport because they make requests repeatable:
 * panning around a neighbourhood re-asks for the same handful of tile keys, so
 * an in-memory cache — and the HTTP cache under it — absorbs almost every
 * movement. The size keeps a tile within the API's 0.05° span cap while
 * covering a marker-zoom viewport in at most a few tiles.
 */

export const TILE_SIZE_DEGREES = 0.02;

export type ViewportRegion = {
  latitude: number;
  longitude: number;
  latitudeDelta: number;
  longitudeDelta: number;
};

export type TileBounds = {
  minLatitude: number;
  maxLatitude: number;
  minLongitude: number;
  maxLongitude: number;
};

/** The tiles covering a region, as stable keys the cache can index by. */
export function tilesForRegion(region: ViewportRegion): string[] {
  const south = region.latitude - region.latitudeDelta / 2;
  const north = region.latitude + region.latitudeDelta / 2;
  const west = region.longitude - region.longitudeDelta / 2;
  const east = region.longitude + region.longitudeDelta / 2;

  const firstRow = Math.floor(south / TILE_SIZE_DEGREES);
  const lastRow = Math.floor(north / TILE_SIZE_DEGREES);
  const firstColumn = Math.floor(west / TILE_SIZE_DEGREES);
  const lastColumn = Math.floor(east / TILE_SIZE_DEGREES);

  const keys: string[] = [];
  for (let row = firstRow; row <= lastRow; row += 1) {
    for (let column = firstColumn; column <= lastColumn; column += 1) {
      keys.push(`${row}:${column}`);
    }
  }
  return keys;
}

/** A tile key back to the bounding box the API expects. */
export function tileBounds(key: string): TileBounds {
  const [row, column] = key.split(':').map(Number);
  return {
    minLatitude: row * TILE_SIZE_DEGREES,
    maxLatitude: (row + 1) * TILE_SIZE_DEGREES,
    minLongitude: column * TILE_SIZE_DEGREES,
    maxLongitude: (column + 1) * TILE_SIZE_DEGREES,
  };
}

File: apps/ios/Via/Domain/ViewportTiles.swift
import Foundation

struct ViewportRegion: Hashable, Sendable {
    let latitude: Double
    let longitude: Double
    let latitudeDelta: Double
    let longitudeDelta: Double
}

enum ViewportTiles {
    static let tileSizeDegrees = 0.02

    static func keys(for region: ViewportRegion) -> [String] {
        let south = region.latitude - region.latitudeDelta / 2
        let north = region.latitude + region.latitudeDelta / 2
        let west = region.longitude - region.longitudeDelta / 2
        let east = region.longitude + region.longitudeDelta / 2

        let firstRow = Int(floor(south / tileSizeDegrees))
        let lastRow = Int(floor(north / tileSizeDegrees))
        let firstColumn = Int(floor(west / tileSizeDegrees))
        let lastColumn = Int(floor(east / tileSizeDegrees))

        return (firstRow...lastRow).flatMap { row in
            (firstColumn...lastColumn).map { column in
                "\(row):\(column)"
            }
        }
    }

    static func bounds(for key: String) -> TileBounds? {
        let parts = key.split(separator: ":")
        guard parts.count == 2, let row = Int(parts[0]), let column = Int(parts[1]) else {
            return nil
        }

        return TileBounds(
            minLatitude: Double(row) * tileSizeDegrees,
            maxLatitude: Double(row + 1) * tileSizeDegrees,
            minLongitude: Double(column) * tileSizeDegrees,
            maxLongitude: Double(column + 1) * tileSizeDegrees
        )
    }
}
